core/server/master/src/main/java/alluxio/master/file/meta/InvalidationSyncCache.java

```diff
@@ -0,0 +1,413 @@
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.master.file.meta;
+
+import alluxio.AlluxioURI;
+import alluxio.collections.Pair;
+import alluxio.conf.Configuration;
+import alluxio.conf.PropertyKey;
+import alluxio.exception.InvalidPathException;
+import alluxio.file.options.DescendantType;
+import alluxio.resource.LockResource;
+import alluxio.util.io.PathUtils;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.time.Clock;
+import java.util.Optional;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.BiConsumer;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * A cache of paths that have been synchronized or have been marked as not being synchronized with
+ * the UFS. Paths are mapped to time values that indicate their last synchronization or
+ * invalidation.
+ *
+ * Synchronization times are updated only at the root path for which a sync was performed by
+ * {@link alluxio.master.file.InodeSyncStream}.
+ * Each path contains the following values concerning synchronization:
+ * - recursiveValidationTime: the last time a sync was successful on this path that included syncing
+ * all of its children recursively
+ * - validationTime: the last time a sync was successful on this path
+ * - directValidationTime: the last time a sync was successful on the path which (at least)
+ * included itself and its direct children
+ * - validationTime: the last time a sync was successful on this path which (at least)
+ * included itself
+ * Each path additionally contains the following values considering invalidation:
+ * - invalidationTime: the last time this exact path was invalidated
+ * - directChildInvalidation: the last time a direct child of this path was invalidated
+ * - recursiveChildInvalidation: the last time a non-direct child of this path was invalidated
+ * - isFile: when checking if a file needs to be synced, any descendant type from the most
+ * recent sync will be valid
+ *
+ * Whenever an invalidation is received the path, and its parents up to the root have their
+ * appropriate invalidation times updated. Validation times are updated on the path
+ * when {@link #notifySyncedPath} is called on the root sync path after a successful sync.
+ *
+ * Checking if a path needs to be synchronized involves checking the appropriate validation
+ * and invalidation times at each path component up to the root.
+ */
+@ThreadSafe
+public class InvalidationSyncCache {
```

Comment:
btw why cannot we keep calling it UFSSyncPathCache? seems like the functionality of this class doesn't change a lot.

Reply:
Another thing is i wonder if we can keep the history of `UfsSyncPathCache`. So we don't add this totally new file but do the change based on rename and changing contents. Otherwise in 3 months you will be asked all question regarding this class, whether a piece of code is really from you or not...

Reply:
Sure, I changed it back.